File: Data Structures & Algorithms/trapping-rain-water/submission-0.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int arrsize = height.size();
        if (arrsize == 0) return 0;
        vector<int> leftmax(arrsize, 0);
        vector<int> rightmax(arrsize, 0);

        leftmax[0] = height[0];
        for (int i=1; i<arrsize; i++) {
            leftmax[i] = std::max(height[i], leftmax[i-1]);
        }

        rightmax[arrsize-1] = height[arrsize-1];
        for (int i=arrsize-2; i>-1; i--) {
            rightmax[i] = std::max(height[i], rightmax[i+1]);
        }

        int total = 0;
        for (int i=0; i<arrsize; i++) {
            int curvol = std::min(leftmax[i], rightmax[i]) - height[i];
            if (curvol > 0 ) total += curvol;
        }
        return total;
    }
};

File: Data Structures & Algorithms/max-water-container/submission-0.py
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        l = 0 
        r = len(heights) - 1
        maxvol = 0
        while l<r:
            vol = (r-l) * min(heights[l], heights[r])
            maxvol = max(maxvol, vol)
            if heights[l] > heights[r]:
                r-=1
            else:
                l+=1
        return maxvol